### ISSUE TITLE
Make lang-items private

### ISSUE THREAD

**Issue by doctorn** [rust-lang/compiler-team#301]
# TL;DR

`tcx.lang_items().my_lang_item().unwrap()` is an anti-pattern which triggers ICEs in `#![no_core]`. This MCP proposes to refactor the current implementation of lang-items to prevent it.

# Links and Details

In `#![no_core]`, we cannot assume that any lang-items are present. At present, due to 

- https://github.com/rust-lang/rust/pull/72170
- https://github.com/rust-lang/rust/pull/72216

the compiler _should_ be free from these kind of ICEs. However, `tcx.lang_items().my_lang_item().unwrap()` is an anti-pattern that we want to avoid reintroducing in the future. Following [some discussion](https://github.com/rust-lang/rust/issues/72240) the idea (credit to @ecstatic-morse) is to introduce an opaque option representing the presence (or lack thereof) of lang-items:

```rust
pub enum LangItemRecord {
    DefId(DefId),
    Missing(LangItem),
}
```

To avoid having to explicitly match on this `enum` everywhere where we normally do something like

```rust
if Some(def_id) == tcx.lang_items().my_lang_item() {
    // ...
}
```

methods would be implemented for `LangItemRecord` to allow for comparison with `DefId`s (silently returning `false` if the item is missing).

By including the `Missing` variant, I am hoping that the `missing` and `items` field of the [`LanguageItems` struct](https://github.com/rust-lang/rust/blob/2a5cbb0e42f330617a061243f8725d861dd5118b/src/librustc_hir/lang_items.rs#L50-L57) can be unified and we can pre-allocate an array of `LangItemRecord`s with each initially assumed `Missing`. This is likely a longer term proposal though and the immediate priority is to prevent `unwrap`s.

The `rustc-dev-guide` is also sadly lacking with respect to descriptions of the various kinds of lang-items and how they're used so I'll be working on supporting documentation as part of this work.

# Mentors or Reviewers

@oli-obk
@ecstatic-morse 
(Assuming both of you are still interested 😅)

**Comment by oli-obk** [comment]
Jup, I'll gladly do any reviews on this

**Comment by nagisa** [comment]
While I'm supportive of the idea, I'd like to point out `#![no_core]` IIRC is unstable and is effectively a `libcore-exclusive` thing anyway, so any ICEs arising as part of using `#![no_core]` are either people using an unstable feature "wrong" or a bug in the rust-lang/rust repo.

(Consequentially, its fine if we adjust the language item design further. I remember @eddyb wanting to change something about how intrinsics/language items worked).

**Comment by bjorn3** [comment]
> so any ICEs arising as part of using #![no_core] are either people using an unstable feature "wrong"

Indeed. For cg_clif I have a crate called `mini_core` that I use in place of `libcore` for some tests as it compiles much faster and when I started `libcore` didn't compile at all. Sometimes when implementing new features in cg_clif I have to expand `mini_core`. It has happened several times that I declared a lang item wrong and the compiler gave no clue about what the problem was.

**Comment by nikomatsakis** [comment]
@bjorn3 to be clear, I also have a "mini core" that I use for debugging purposes. I don't mean that we shouldn't have the capability, but I do mean that we shouldn't expose it on stable, and I think that we should feel free to make changes to libcore impl. details and lang item definitions liberally.
